### PR TITLE
Add better support for different forge web URLs, flatten BaseBranch f…

### PR DESCRIPTION
**Pull request by Hellebore**
…rom repeat logic.

## ☕️ Reasoning
- Support for Azure DevOps branch and commit URLs
- Easier future support for more forge types without deep integrations

## 🧢 Changes

- Moved forge type from a collection of Boolean functions to their own type, new Azure DevOps, and a default Unknown.
- Added test cases for each forge type.
- Flattened `BaseBranch` logic to do potentially repetitive work once, most of which could be simplified further if moved to the rust side still.
- Moved some logic around to better support performing string actions once, namely `convertRemoteToWebUrl` and tests there of.

## Summary by Sourcery

This pull request enhances the support for different forge web URLs, including Azure DevOps, by refactoring the forge type handling and simplifying the BaseBranch logic. It also includes comprehensive tests to ensure the correctness of the new functionality.

- **New Features**:
    - Added support for Azure DevOps branch and commit URLs.
- **Enhancements**:
    - Refactored forge type handling by introducing a new Forge enum and associated type.
    - Simplified BaseBranch logic by consolidating repetitive operations and moving some logic to Rust.
    - Improved URL handling by removing the convertRemoteToWebUrl function and integrating GitUrlParse for URL parsing.
- **Tests**:
    - Added test cases for each forge type to ensure correct URL parsing and commit URL generation.
    - Introduced tests to validate the handling of various remote URLs, including edge cases like SSH URLs with IP addresses.

### CODE REVIEW

**Comment by Hellebore** [comment]
This is a benchmark review for experiment `gitbutler`.
Run ID: `gitbutler/benchmark_2024-07-19T09-41-28_v1-21-0-26-g63c6b8476`.

This pull request was cloned from `https://github.com/gitbutlerapp/gitbutler/pull/4426`. _(Note: the URL is not a link to avoid triggering a notification on the original pull request.)_
<details>
<summary>Experiment configuration</summary>

```yaml
review_config:
  # User configuration for the review
  # - benchmark - use the user config from the benchmark reviews
  # - <value> - use the value directly
  user_review_config:
    enable_ai_review: true
    enable_rule_comments: false

    enable_complexity_comments: benchmark
    enable_security_comments: benchmark
    enable_tests_comments: benchmark
    enable_comment_suggestions: benchmark
    enable_functionality_review: benchmark

    enable_pull_request_summary: benchmark
    enable_review_guide: benchmark

    enable_approvals: true

  ai_review_config:
    # The model responses to use for the experiment
    # - benchmark - use the model responses from the benchmark reviews
    # - llm - call the language model to generate responses
    model_responses:
      comments_model: benchmark
      comment_area_model: benchmark
      comment_validation_model: benchmark
      comment_suggestion_model: benchmark
      complexity_model: benchmark
      docstrings_model: benchmark
      functionality_model: benchmark
      security_model: benchmark
      tests_model: benchmark
      pull_request_summary_model: benchmark
      review_guide_model: benchmark

# The pull request dataset to run the experiment on
pull_request_dataset:
- https://github.com/gitbutlerapp/gitbutler/pull/4426
- https://github.com/gitbutlerapp/gitbutler/pull/4224
- https://github.com/gitbutlerapp/gitbutler/pull/4069
- https://github.com/gitbutlerapp/gitbutler/pull/3386
- https://github.com/gitbutlerapp/gitbutler/pull/3719

```

</details>

**Comment by SourceryAI** [comment]
## Reviewer's Guide by Sourcery

This pull request refactors the BaseBranch class to support different forge types (GitHub, GitLab, Bitbucket, AzureDevOps) by introducing a new Forge enum and related properties. It also moves URL conversion logic into the BaseBranch class and adds post-construction processing via the afterTransform method. Additionally, it updates the BaseBranchSwitch component and related test cases to align with these changes.

### File-Level Changes

| Files | Changes |
| ----- | ------- |
| `app/src/lib/vbranches/types.ts`<br/>`app/src/lib/vbranches/baseBranch.ts` | Refactored BaseBranch class to support different forge types and added post-construction logic. |
| `app/src/lib/utils/url.ts`<br/>`app/src/lib/utils/url.test.ts` | Removed convertRemoteToWebUrl function and updated related test cases. |

---

<details>
<summary>Tips</summary>

- Trigger a new Sourcery review by commenting `@sourcery-ai review` on the pull request.
- Continue your discussion with Sourcery by replying directly to review comments.
- You can change your review settings at any time by accessing your [dashboard](https://app.sourcery.ai):
    - Enable or disable the Sourcery-generated pull request summary or reviewer's guide;
    - Change the review language;
- You can always [contact us](mailto:support@sourcery.ai) if you have any questions or feedback.

</details>

File: app/src/lib/vbranches/types.ts [hunk 1]
```diff
@@ -370,12 +371,20 @@ export class RemoteBranchData {
 	}
 }
 
+export enum Forge {
+	Unknown,
+	GitHub,
+	GitLab,
+	Bitbucket,
+	AzureDevOps
+}
```
Comment (SourceryAI) [inline app/src/lib/vbranches/types.ts:380]: **suggestion:** Consider adding a description for each enum value.

Adding a description for each enum value can improve code readability and maintainability, especially for new developers or when the codebase grows.

Comment (SourceryAI) [inline app/src/lib/vbranches/types.ts:380]: **suggestion:** Consider using a map for forge type determination.

Using a map for forge type determination can make the code more concise and potentially faster.

File: app/src/lib/vbranches/baseBranch.ts [hunk 2]
```diff
@@ -104,7 +104,15 @@ export class BaseBranchService {
 }
 
 async function getBaseBranch(params: { projectId: string }): Promise<BaseBranch | null> {
-	return plainToInstance(BaseBranch, await invoke<any>('get_base_branch_data', params));
+	return createBaseBranch(await invoke<any>('get_base_branch_data', params));
+}
```
Comment (SourceryAI) [inline app/src/lib/vbranches/baseBranch.ts:108]: **suggestion:** Consider adding type validation for the input object.

Adding type validation for the input object can prevent potential runtime errors and improve code robustness.

File: app/src/lib/vbranches/baseBranch.ts [hunk 3]
```diff
@@ -104,7 +104,15 @@ export class BaseBranchService {
 }
 
 async function getBaseBranch(params: { projectId: string }): Promise<BaseBranch | null> {
-	return plainToInstance(BaseBranch, await invoke<any>('get_base_branch_data', params));
+	return createBaseBranch(await invoke<any>('get_base_branch_data', params));
```
Comment (SourceryAI) [inline app/src/lib/vbranches/baseBranch.ts:107]: **issue (complexity):** Consider simplifying the code by reducing the use of `rxjs` operators.

This code is quite complex due to several factors:

1. **Increased Number of Dependencies**: The code imports a large number of modules and functions from various libraries (`rxjs`, `class-transformer`, etc.), which increases the cognitive load on anyone trying to understand it.

2. **Complex Observable Chains**: The use of `rxjs` operators like `combineLatest`, `mergeWith`, `debounceTime`, `switchMap`, `tap`, `catchError`, `startWith`, and `shareReplay` creates a complex chain of operations, making the code harder to read and understand, especially for developers not well-versed in reactive programming.

3. **Error Handling**: Error handling is spread across multiple places (`catchError` in the observable chain, `try-catch` blocks in methods), making it difficult to trace the flow of errors and understand how they are being handled.

4. **State Management**: The use of multiple `BehaviorSubject` and `Subject` instances for managing state (`busy$`, `error$`, `reload$`) adds to the complexity, making it harder to reason about the state of the application at any given time.

5. **Asynchronous Operations**: The code contains several asynchronous operations (`invoke`, `getBaseBranch`, `fetchFromRemotes`, `setTarget`). Managing these operations and ensuring they complete in the correct order adds to the complexity.

Consider simplifying the code by reducing the use of `rxjs` operators and focusing on a more straightforward approach. For example, using Svelte's `writable` store for state management and centralizing error handling within `try-catch` blocks can make the code easier to read, understand, and maintain.

**Review by SourceryAI — APPROVED** [review]
Hey @Hellebore - I've reviewed your changes and they look great!

<details>
<summary>Here's what I looked at during the review</summary>

- 🟡 **General issues**: 3 issues found
- 🟢 **Security**: all looks good
- 🟢 **Testing**: all looks good
- 🟡 **Complexity**: 1 issue found
- 🟢 **Documentation**: all looks good
</details>

***

[LangSmith trace](https://smith.langchain.com/o/5c8bea2b-6c6c-4221-bd9b-0c11556933a8/projects/p/2dbf7697-3bb3-4f63-b184-3a8ebf7ac23c/r/8febb9e3-4a1e-4de9-ad29-310f7311282c?poll=true)

<sub>
Help me be more useful! Please click 👍 or 👎 on each comment to tell me if it was helpful.
</sub>
